### PR TITLE
Add non-allocating helper function for torch.linalg.qr

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57317 Add cuSOLVER path for torch.linalg.lstsq
* #57316 Add CUDA support for torch.ormqr
* #57315 Port CPU torch.ormqr to ATen
* #57314 Fix torch.ormqr for non Fortran-contiguous inputs
* #56257 Fix MAGMA qr for empty batched inputs
* #56256 Add cuSOLVER path for torch.linalg.qr
* #56255 Remove size arguments for internal orgqr and geqrf calls
* **#56254 Add non-allocating helper function for torch.linalg.qr**

Differential Revision: [D27960151](https://our.internmc.facebook.com/intern/diff/D27960151)

### CODE REVIEW

**Comment by mruberry** [comment]
@Lezcano would you sanity check this?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -1693,55 +1693,119 @@ std::tuple<Tensor, Tensor> geqrf(const Tensor& input) {
...
+  auto mn = std::min(m, n);
+
+  // Q must have the expected shape: reduced_mode ? (..., m, min(m, n)) : (..., m, m)
+  if (compute_q) {
+    auto expected_Q_shape = input.sizes().vec();
+    expected_Q_shape.back() = reduced_mode ? mn : m;
+    TORCH_INTERNAL_ASSERT(Q.sizes().equals(expected_Q_shape));
+
+    // Q tensor must be in batched column major order (Fortran contiguous)
+    TORCH_INTERNAL_ASSERT(Q.transpose(-2, -1).is_contiguous());
```
Comment (antocuni) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1729]: this has nothing to do with this PR, it's just a random idea which came to my mind while reading this.
Should we have something like `is_fortran_contiguous()`? And maybe even introduce `at::MemoryFormat::FortranContiguous` which would be useful when creating tensors.
I have seen similar patterns a bit everywhere in linalg code, and having first-class support for fortran-style memory would probably simplify a lot of code

**Review by antocuni — COMMENTED** [review]
LGTM.
If I understand it correctly, this PR by itself is mostly a no-op because it introduces `_linalg_qr_out_helper` but it never uses it apart for re-implementing the old `_linalg_qr_helper`, but I assume it will be used by subsequent PRs in the ghstack

**Review by lezcano — COMMENTED** [review]
It looks very good. I just found one part that was a bit difficult to follow. I made a proposal for a simpler way to handle that, which I think should be save one allocation.

If this sounds reasonable, the handling of the "unpacking" of the tensors after the `geqrf` call should be changed as well.

**Comment by mruberry** [comment]
@Lezcano would you take a look at this PR?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:1729]: In linalg code we need to have only that the last two dimensions to be Fortran contiguous and all other are normal C contiguous. Real Fortran-style memory would be to have the matrix in the first two dimensions and batches in the end, it's not useful to have.

**Comment by IvanYashchuk** [comment]
> If I understand it correctly, this PR by itself is mostly a no-op because it introduces _linalg_qr_out_helper but it never uses it apart for re-implementing the old _linalg_qr_helper, but I assume it will be used by subsequent PRs in the ghstack

The previous helper was allocating memory on the go, with narrowing and resizes, which is not a good strategy to do. Here we first allocate the memory of the correct size and then fill it with the result.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -1693,55 +1693,119 @@ std::tuple<Tensor, Tensor> geqrf(const Tensor& input) {
...
+  auto tau_shape = input.sizes().vec();
+  tau_shape.pop_back();
+  tau_shape.back() = mn;
+  Tensor tau = at::empty(tau_shape, input.options());
 
-  geqrf_stub(q_working_copy.device().type(), q_working_copy, tau_working_copy, m, n);
+  // geqrf requires m x n workspace input that is modified in-place
+  // if m > n and reduced==true we use Q tensor for storing the result of geqrf operation
+  // otherwise R tensor is used
+  Tensor QR = R;
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:1748]: > the current implementation allocates a new tensor whenever compute_q == false, which is not necessary

Right, I'll fix that.

[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: I changed the code to use your conditions.

**Review by lezcano — APPROVED** [review]
LGTM!

**Review by mruberry — APPROVED** [review]
Stamped

**Pull request closed without merging**